### DISCUSSION TITLE
Solution for commentators. Ingest RTMP and display via WEBRTC. Comment via WebRTC and mix it to hls

### DISCUSSION THREAD

**Discussion by VitalyZaikin** [OvenMediaLabs/OvenMediaEngine#1743] [Help] [answered]
Hi there! I'm new here.
I need to get an RTMP stream from the studio and display it via WebRTC. The commentator will view this stream via WebRTC. Through WebRTC, I send audio to OME. Inside OME, I transcode the audio to AAC and push it via RTMP. The main idea to get HLS with audi track switching 
Here is my config. I would appreciate your help. 

?xml version="1.0" encoding="UTF-8" ?>

<Server version="8">
	<Name>OvenMediaEngine</Name>
	
	<Type>origin</Type>
	
	<IP>*</IP>
	<PrivacyProtection>false</PrivacyProtection>
	
	<StunServer>stun.ovenmediaengine.com:13478</StunServer>

	<Modules>
		
		<HTTP2>
			<Enable>true</Enable>
		</HTTP2>

		<LLHLS>
			<Enable>true</Enable>
		</LLHLS>

		
		<P2P>
			
			<Enable>false</Enable>
			<MaxClientPeersPerHostPeer>2</MaxClientPeersPerHostPeer>
		</P2P>
	</Modules>

	
	<Bind>
		
		
		<Providers>
			
			<RTSPC>
				<WorkerCount>1</WorkerCount>
			</RTSPC>
			<OVT>
				<WorkerCount>1</WorkerCount>
			</OVT>
			
			<RTMP>
				<Port>${env:OME_RTMP_PROV_PORT:1935}</Port>
				<WorkerCount>1</WorkerCount>
			</RTMP>
			<SRT>
				<Port>${env:OME_SRT_PROV_PORT:9999}</Port>
				<WorkerCount>1</WorkerCount>
			</SRT>
			
			<WebRTC>
				<Signalling>
					<Port>${env:OME_WEBRTC_SIGNALLING_PORT:3333}</Port>
					<TLSPort>${env:OME_WEBRTC_SIGNALLING_TLS_PORT:3334}</TLSPort>
					<WorkerCount>1</WorkerCount>
				</Signalling>
				<IceCandidates>
					<IceCandidate>${env:OME_HOST_IP:*}:${env:OME_WEBRTC_CANDIDATE_PORT:10000-10004/udp}</IceCandidate>
					<TcpRelay>${env:OME_HOST_IP:*}:${env:OME_WEBRTC_TCP_RELAY_PORT:3478}</TcpRelay>
					<TcpForce>true</TcpForce>
					<TcpRelayWorkerCount>1</TcpRelayWorkerCount>
				</IceCandidates>
			</WebRTC>
		</Providers>

		<Publishers>
			
			<OVT>
				<Port>${env:OME_ORIGIN_PORT:9000}</Port>
				<WorkerCount>1</WorkerCount>
			</OVT>
			<LLHLS>
				
				<Port>${env:OME_LLHLS_STREAM_PORT:3333}</Port>
				<TLSPort>${env:OME_LLHLS_STREAM_TLS_PORT:3334}</TLSPort>
				<WorkerCount>1</WorkerCount>
			</LLHLS>
			<WebRTC>
				<Signalling>
					<Port>${env:OME_WEBRTC_SIGNALLING_PORT:3333}</Port>
					<TLSPort>${env:OME_WEBRTC_SIGNALLING_TLS_PORT:3334}</TLSPort>
					<WorkerCount>1</WorkerCount>
				</Signalling>
				<IceCandidates>
					<IceCandidate>${env:OME_HOST_IP:*}:${env:OME_WEBRTC_CANDIDATE_PORT:10000-10004/udp}</IceCandidate>
					<TcpRelay>${env:OME_HOST_IP:*}:${env:OME_WEBRTC_TCP_RELAY_PORT:3478}</TcpRelay>
					<TcpForce>true</TcpForce>
					<TcpRelayWorkerCount>1</TcpRelayWorkerCount>
				</IceCandidates>
			</WebRTC>
		</Publishers>
	</Bind>

	<VirtualHosts>
		
		<VirtualHost include="VHost*.xml" />
		<VirtualHost>
			<Name>default</Name>
			
			<Distribution>ovenmediaengine.com</Distribution>
			
			
			<Host>
				<Names>
					
					<Name>*</Name>
				</Names>
				<TLS>
					<CertPath>./cert.crt</CertPath>
					<KeyPath>./cert.key</KeyPath>
					<ChainCertPath>./cert.ca-bundle</ChainCertPath>
				</TLS>
			</Host>

			

			

			
			<CrossDomains>
				<Url>*</Url>
			</CrossDomains>

			
			<Applications>
			<Application>
    <Name>live</Name>
    <Type>live</Type>
    <Streams>
        <Stream>
            <Name>main_video_stream</Name>
            <Input>
                <WebRTC>
                    <Url>wss://streamihive.live:3334/app/stream</Url>
                </WebRTC>
            </Input>
            <Outputs>
                <SelectOutputTrack>video</SelectOutputTrack>
            </Outputs>
        </Stream>

        <Stream>
            <Name>first_audio_stream</Name>
            <Input>
                <WebRTC>
                    <Url>wss://streamihive.live:3334/app/stream</Url>
                </WebRTC>
            </Input>
            <Outputs>
                <Transcode>
                    <Audio>
                        <Codec>aac</Codec>
                        <Bitrate>128000</Bitrate>
                        <Samplerate>44100</Samplerate>
                        <Channel>2</Channel>
                    </Audio>
                </Transcode>
            </Outputs>
        </Stream>

        <Stream>
            <Name>second_audio_stream</Name>
            <Input>
                <WebRTC>
                    <Url>wss://streamihive.live:3334/app/stream2</Url>
                </WebRTC>
            </Input>
            <Outputs>
                <Transcode>
                    <Audio>
                        <Codec>aac</Codec>
                        <Bitrate>128000</Bitrate>
                        <Samplerate>44100</Samplerate>
                        <Channel>2</Channel>
                    </Audio>
                </Transcode>
            </Outputs>
        </Stream>

        <Stream>
            <Name>mixed_stream</Name>
            <Input>
                <Stream>main_video_stream</Stream>
                <Stream>first_audio_stream</Stream>
                <Stream>second_audio_stream</Stream>
            </Input>
            <Outputs>
                <RTMP>
                    <Url>rtmp://streamihive.live/live/mixed_stream</Url>
                    <MultiAudio>
                        <Track1>first_audio_stream</Track1>
                        <Track2>second_audio_stream</Track2>
                    </MultiAudio>
                </RTMP>
            </Outputs>
        </Stream>
    </Streams>
</Application>

				<Application>
					<Name>app</Name>
					
					<Type>live</Type>
					<OutputProfiles>
						
						

						
						<HWAccels>
							<Decoder>
									<Enable>false</Enable>
									
									
							</Decoder>
							<Encoder>
									<Enable>false</Enable>
									
							</Encoder>
						</HWAccels>

						<OutputProfile>
							<Name>bypass_stream</Name>
							<OutputStreamName>${OriginStreamName}</OutputStreamName>

							

							<Encodes>
								<Video>
									<Name>bypass_video</Name>
									<Bypass>true</Bypass>
								</Video>
								<Audio>
									<Name>aac_audio</Name>
									<Codec>aac</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
									<BypassIfMatch>
										<Codec>eq</Codec>
									</BypassIfMatch>
								</Audio>								
								<Audio>
									<Name>opus_audio</Name>
									<Codec>opus</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
									<BypassIfMatch>
										<Codec>eq</Codec>
									</BypassIfMatch>									
								</Audio>
								
							</Encodes>
						</OutputProfile>
					</OutputProfiles>
					<Providers>
						<OVT />
						<WebRTC>
							<Timeout>30000</Timeout>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
						</WebRTC>
						<RTMP />
						<SRT />
						<RTSPPull />
						<Schedule>
							<MediaRootDir>/opt/ovenmediaengine/media</MediaRootDir>
							<ScheduleFilesDir>/opt/ovenmediaengine/media</ScheduleFilesDir>
						</Schedule>
						
					</Providers>
					<Publishers>
						<AppWorkerCount>1</AppWorkerCount>
						<StreamWorkerCount>8</StreamWorkerCount>
						<OVT />
						<WebRTC>
							<Timeout>30000</Timeout>
							<Rtx>false</Rtx>
							<Ulpfec>false</Ulpfec>
							<JitterBuffer>false</JitterBuffer>
							<CreateDefaultPlaylist>true</CreateDefaultPlaylist>
						</WebRTC>
						<LLHLS>
							<OriginMode>true</OriginMode>
							
							<ChunkDuration>0.5</ChunkDuration>
							
							<PartHoldBack>1.5</PartHoldBack>
							<SegmentDuration>6</SegmentDuration>
							<SegmentCount>10</SegmentCount>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
							<CreateDefaultPlaylist>true</CreateDefaultPlaylist> 
						</LLHLS>
						<HLS>
							<SegmentCount>4</SegmentCount>
							<SegmentDuration>4</SegmentDuration>
							
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
							<CreateDefaultPlaylist>true</CreateDefaultPlaylist> 
						</HLS>
					</Publishers>
				</Application>
			</Applications>
		</VirtualHost>
	</VirtualHosts>
</Server>

**Comment by nums** [comment]
You can take a look here : https://airensoft.gitbook.io/ovenmediaengine/live-source/multiplex-channel

**Reply by VitalyZaikin** [reply]
@nums Thanks for the answer. It still doesn't work for me, don't know why. Could you share some examples of worked config? 

Here is my Server config


<Providers>
						<OVT />
						<WebRTC>
							<Timeout>30000</Timeout>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
						</WebRTC>
						<RTMP />
						<SRT />
						<RTSPPull />
						<Schedule>
							<MediaRootDir>/opt/ovenmediaengine/media</MediaRootDir>
							<ScheduleFilesDir>/opt/ovenmediaengine/media</ScheduleFilesDir>
						</Schedule>
						<Multiplex>
			<MuxFilesDir>/usr/share/ovenmediaengine/conf/</MuxFilesDir>
		</Multiplex>


And here is mux.mux file 
<?xml version="1.0" encoding="UTF-8"?>
<Multiplex>
    <OutputStream>
        <Name>mux</Name>
    </OutputStream>

    <SourceStreams>
        <SourceStream>
            <Name>tv1</Name>
            <Url>wss://streamihive.live:3334/app/stream2</Url>
            <TrackMap>
                <Track>
                    <SourceTrackName>bypass_video</SourceTrackName>
                    <NewTrackName>tv1_video</NewTrackName>
                </Track>
                <Track>
                    <SourceTrackName>bypass_audio</SourceTrackName>
                    <NewTrackName>tv1_audio</NewTrackName>
                </Track>
                <Track>
                    <SourceTrackName>opus</SourceTrackName>
                    <NewTrackName>tv1_opus</NewTrackName>
                </Track>
            </TrackMap>
        </SourceStream>
        <SourceStream>
            <Name>tv2</Name>
            <Url>wss://streamihive.live:3334/app/stream2</Url>

            <TrackMap>
                <Track>
                    <SourceTrackName>bypass_video</SourceTrackName>
                    <NewTrackName>tv2_video</NewTrackName>
                </Track>
                <Track>
                    <SourceTrackName>bypass_audio</SourceTrackName>
                    <NewTrackName>tv2_audio</NewTrackName>
                </Track>
                <Track>
                    <SourceTrackName>opus</SourceTrackName>
                    <NewTrackName>tv2_opus</NewTrackName>
                </Track>
            </TrackMap>

        </SourceStream>
    </SourceStreams>
    
    <Playlists>
        <Playlist>
            <Name>123</Name>
            <FileName>abr</FileName>
            
            <Rendition>
                <Name>1080p</Name>
                <Video>tv1_video</Video>
                <Audio>tv1_audio</Audio>
            </Rendition>
            <Rendition>
                <Name>720p</Name>
                <Video>tv2_video</Video>
                <Audio>tv2_audio</Audio>
            </Rendition>
        </Playlist>
    </Playlists>
</Multiplex>

I really can't recognize where are the issues

**Reply by getroot** [reply, marked as answer]
Mux Channel is a feature that allows you to take a stream created inside OME and reassemble the tracks to create a new channel. You cannot enter a URL starting with wss://, and according to the manual, enter information about a stream already created with WebRTC according to the stream://vhost_name/app_name/stream_name rule.

**Reply by VitalyZaikin** [reply]
Thanks, was a misunderstanding about the stream format. Just for understanding, is it possible to mix audio pairs, like mono, stereo?

**Reply by nums** [reply]
For now, there is not mixing feature or channel audio track management but you can provide 2 HLS streams : one with video studio + audio studio and the other with video studio + audio commentators

**Reply by VitalyZaikin** [reply]
> For now, there is not mixing feature or channel audio track management but you can provide 2 HLS streams : one with video studio + audio studio and the other with video studio + audio commentators

Thanks, really helpful. One more question, does webtrtc stream streaming in stereo, or it's mono in 2 pairs?
